src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -964,50 +956,33 @@ internal static void CreateNameToMembersMap
         {
             var dictionary = new Dictionary<string, ImmutableArray<TNamedTypeSymbol>>(comparer);
 
-            foreach (var kvp in map)
-            {
-                ImmutableArray<TNamespaceOrTypeSymbol> members = kvp.Value;
+            foreach (var (name, members) in map)
+                dictionary.Add(name, getOrCreateNamedTypes(members));
 
-                bool hasType = false;
-                bool hasNamespace = false;
+            return dictionary;
 
-                foreach (var symbol in members)
+            ImmutableArray<TNamedTypeSymbol> getOrCreateNamedTypes(ImmutableArray<TNamespaceOrTypeSymbol> members)
+            {
+                // See if creator 'map' put a downcasted ImmutableArray<TNamedTypeSymbol> in it.  If so, we can just directly
+                // downcast to that and trivially reuse it.  If not, that means the array must have contained at least one
+                // TNamespaceSymbol and we'll need to filter that out.
+                var membersAsNamedTypes = members.As<TNamedTypeSymbol>();
+
+                if (!membersAsNamedTypes.IsDefault)
+                    return membersAsNamedTypes;
+
+                // Preallocate the right amount so we can avoid garbage reallocs.
+                var count = members.Count(static s => s is TNamedTypeSymbol);
+                var builder = ArrayBuilder<TNamedTypeSymbol>.GetInstance(count);
+                foreach (var member in members)
                 {
-                    if (symbol is TNamedTypeSymbol)
-                    {
-                        hasType = true;
-                        if (hasNamespace)
-                        {
-                            break;
-                        }
-                    }
-                    else
-                    {
-#if DEBUG
-                        Debug.Assert(symbol is TNamespaceSymbol);
-#endif
-                        hasNamespace = true;
-                        if (hasType)
-                        {
-                            break;
-                        }
-                    }
+                    if (member is TNamedTypeSymbol namedType)
+                        builder.Add(namedType);
                 }
 
-                if (hasType)
-                {
-                    if (hasNamespace)
-                    {
-                        dictionary.Add(kvp.Key, members.OfType<TNamedTypeSymbol>().AsImmutable());
-                    }
-                    else
-                    {
-                        dictionary.Add(kvp.Key, members.As<TNamedTypeSymbol>());
-                    }
-                }
+                Debug.Assert(builder.Count == count);
+                return builder.ToImmutableAndFree();
```

Comment:
>return builder.ToImmutableAndFree(); [](http://example.com/codeflow?start=16&length=36)

It looks like we will be adding empty entries into the resulting dictionary, this isn't what we want and isn't what we did before.

Reply:
makes sense.  revised.